extensions/ext/GL_EXT_subgroupuniform_qualifier.txt

```diff
@@ -0,0 +1,158 @@
+    John Kessenich, Google
+
+Status
+
+    Draft
+
+Version
+
+    Last Modified Date:         September 19, 2018
+    Revision:                   1
+
+Number
+
+    TBD
+
+Dependencies
+
+    This extension requires GL_KHR_vulkan_glsl
+
+Overview
+
+    This extension adds a "subgroupuniformEXT" type qualifier and constructor,
+    such that for each dynamic instance of an object qualified by a subgroup
+    uniform type or constructed with a subgroup uniform constructor, all active
+    invocations in a subgroup compute the same result value.
+
+New Procedures and Functions
+
+    None.
+
+New Tokens
+
+    None.
+
+Modifications to GL_KHR_vulkan_glsl
+
+    Add to the "Mapping to SPIR-V" section
+    
+    Mapping of subgroupuniformEXT type qualifier:
+
+      subgroupuniformEXT -> subgroupuniformEXT decoration on variables
+
+Modifications to the OpenGL Shading Language Specification, Version 4.50
+
+    Including the following line in a shader can be used to control the
+    language features described in this extension:
+
+        #extension GL_EXT_subgroupuniform_qualifier : <behavior>
+
+    where <behavior> is as specified in section 3.3
+
+    New preprocessor #defines are added to the OpenGL Shading Language:
+
+        #define GL_EXT_subgroupuniform_qualifier     1
+
+    Add to section 3.6 Keywords:
+
+        subgroupuniformEXT
+
+    Add a new section:
+      
+      "4.X subgroupuniformEXT qualifier"
+
+      The subgroupuniformEXT qualifier can be used to assert that for a dynamic
+      instance of a variable or expression all active invocations in a subgroup
+      compute the same result value. In a declaration, it is syntactically
+      treated as a qualifier. It can be applied to:
+
+        * variable declarations qualified as *in*
+        * global variable declarations with no storage qualifier
+        * local variable declarations with no storage qualifier
+        * function parameter declarations and function return types.
+
+      Any other use on a declaration results in a compile-time error.
+
+      The subgroupuniformEXT qualifier can also be used with constructor syntax
+      to assert that for a dynamic instance of an expression all active
+      invocations in a subgroup compute the same result value. For example:
+
+          if (subgroupuniformEXT(a < b)) {
+             // All active invocations in the subgroup execute this or none will
+          }
+
+      This constructor syntax takes a single argument of any type and returns
+      the value with the same type, qualified with subgroupuniformEXT.
+
+      Only some operations discussed in Chapter 5 (Operators and Expressions)
+      can be applied to subgroup uniform value(s) and still yield a result that
+      is subgroup uniform. The operations that do so are listed below. When a
+      subgroup uniform value is operated on with one of these operators
+      (regardless of whether any and other operands are subgroup uniform), the
```

Comment:
This needs to be changed from how nonuniformEXT worked. For nonuniformEXT, if any operand is nonuniform then the result is nonuniform. For subgroupuniform, we would want to say that the result is subgroupuniform only if all operands are subgroupuniform.

I didn't review the list of operations below, but it might need to be different from nonuniform as well.